test/cpu_reduce_util.hpp

```diff
@@ -29,19 +29,177 @@
 #include <half.hpp>
 #include <limits>
 #include <cmath>
+#include <cassert>
+#include <stdexcept>
+#include <string>
 #include <miopen/miopen.h>
 #include <miopen/reduce_common.hpp>
 
 #include <miopen/bfloat16.hpp>
 
 namespace reduce {
 
+template <typename T>
+static inline bool IsNan(T x)
+{
+    // C++ isnan() is used for float, double and half_float::half
+    return (std::isnan(x));
+};
+
+template <typename T>
+static inline bool float_equal_one(T x)
+{
+    (void)x;
+    static_assert(static_cast<T>(0), "float_equal_one() is not implemented for this data type");
+    return false;
+};
+
+template <>
+inline bool float_equal_one<float>(float x)
+{
+    return x == 1.0f;
+};
+
+template <>
+inline bool float_equal_one<double>(double x)
+{
+    return x == 1.0;
+};
+
+template <>
+inline bool float_equal_one<half_float::half>(half_float::half x)
+{
+    return x == convert_type<half_float::half>(1.0f);
+};
+
+template <typename T>
+static inline bool float_equal_zero(T x)
+{
+    (void)x;
+    static_assert(static_cast<T>(0), "float_equal_zero() is not implemented for this data type");
+    return false;
+};
+
+template <>
+inline bool float_equal_zero<float>(float x)
+{
+    return x == 0.0f;
+};
+
+template <>
+inline bool float_equal_zero<double>(double x)
+{
+    return x == 0.0;
+};
+
+template <>
+inline bool float_equal_zero<half_float::half>(half_float::half x)
+{
+    return x == convert_type<half_float::half>(0.0f);
+};
+
+template <typename T>
+static inline T Sqrt(T a)
+{
+    return sqrt(a);
+};
+
+template <>
+inline float Sqrt<float>(float a)
+{
+    return sqrtf(a);
+};
+
+template <>
+inline half_float::half Sqrt<half_float::half>(half_float::half a)
+{
+    return half_float::sqrt(a);
+};
+
+template <typename T>
+static inline T Abs(T a)
+{
+    return abs(a);
+};
+
+template <>
+inline float Abs<float>(float a)
+{
+    return fabsf(a);
+};
+
+template <>
+inline double Abs<double>(double a)
+{
+    return fabs(a);
+};
+
+template <>
+inline half_float::half Abs<half_float::half>(half_float::half a)
+{
+    return half_float::abs(a);
+};
+
+template <typename compType>
+static inline std::function<void(compType&)> PreUnaryOpFn(miopenReduceTensorOp_t op_, int divider)
```

Comment:
If performance matters then this should use a visitor pattern instead of returning `std::function`:

```cpp
template <class compType, class F>
static inline void PreUnaryOpFn(miopenReduceTensorOp_t op_, int divider, F f)
{
    switch(op_)
    {
    case MIOPEN_REDUCE_TENSOR_NORM1:
        f([&, divider](compType& a_) {
            a_ = Abs<compType>(a_) / convert_type<compType>(static_cast<float>(divider));
        });
    case MIOPEN_REDUCE_TENSOR_NORM2:
        f([&, divider](compType& a_) {
            a_ = a_ * a_ / convert_type<compType>(static_cast<float>(divider));
        });
    case MIOPEN_REDUCE_TENSOR_AMAX: f([&](compType& a_) { a_ = Abs<compType>(a_); });

    case MIOPEN_REDUCE_TENSOR_AVG:
    case MIOPEN_REDUCE_TENSOR_ADD:
    case MIOPEN_REDUCE_TENSOR_MUL:
    case MIOPEN_REDUCE_TENSOR_MIN:
    case MIOPEN_REDUCE_TENSOR_MAX: f([&](compType& a_) { (void)a_; });
    }

    throw std::runtime_error(std::string(__FUNCTION__) +
                             ": using undefined Reduction operation is not permitted");
};
```

Then you would call it like this:

```cpp
PreUnaryOpFn<compType>(reduceOp, divider, [&](auto PreUnaryOp) {
    
});
```

But you probably need to combine both `PreUnaryOpFn` and `PosUnaryOpFn` so it can reuse the same switch statement.

Reply:
I could not very much understand your point here.

Reply:
@fronteer I highly recommend spending some time for learning this topic, at least read https://en.wikipedia.org/wiki/Visitor_pattern